diff --git a/mod/assign/locallib.php b/mod/assign/locallib.php
index 55cb0d5..aaf848c 100644
--- a/mod/assign/locallib.php
+++ b/mod/assign/locallib.php
@@ -4398,6 +4398,7 @@ class assign {
         $eventdata->notification    = 1;
         $eventdata->contexturl      = $info->url;
         $eventdata->contexturlname  = $info->assignment;
+        $eventdata->courseid        = $course->id;
 
         message_send($eventdata);
     }
diff --git a/mod/feedback/lib.php b/mod/feedback/lib.php
index 5a9ed8d..de32ae0 100644
--- a/mod/feedback/lib.php
+++ b/mod/feedback/lib.php
@@ -2944,7 +2944,10 @@ function feedback_send_email($cm, $feedback, $course, $userid) {
                 $eventdata->fullmessage      = $posttext;
                 $eventdata->fullmessageformat = FORMAT_PLAIN;
                 $eventdata->fullmessagehtml  = $posthtml;
-                $eventdata->smallmessage     = '';
+                $eventdata->smallmessage     = $postsubject;
+                $eventdata->courseid         = $course->id;
+                $eventdata->contexturl       = $info->url;
+                $eventdata->contexturlname   = $info->feedback;
                 message_send($eventdata);
             } else {
                 $eventdata = new stdClass();
@@ -2956,7 +2959,10 @@ function feedback_send_email($cm, $feedback, $course, $userid) {
                 $eventdata->fullmessage      = $posttext;
                 $eventdata->fullmessageformat = FORMAT_PLAIN;
                 $eventdata->fullmessagehtml  = $posthtml;
-                $eventdata->smallmessage     = '';
+                $eventdata->smallmessage     = $postsubject;
+                $eventdata->courseid         = $course->id;
+                $eventdata->contexturl       = $info->url;
+                $eventdata->contexturlname   = $info->feedback;
                 message_send($eventdata);
             }
         }
@@ -3013,7 +3019,10 @@ function feedback_send_email_anonym($cm, $feedback, $course) {
             $eventdata->fullmessage      = $posttext;
             $eventdata->fullmessageformat = FORMAT_PLAIN;
             $eventdata->fullmessagehtml  = $posthtml;
-            $eventdata->smallmessage     = '';
+            $eventdata->smallmessage     = $postsubject;
+            $eventdata->courseid         = $course->id;
+            $eventdata->contexturl       = $info->url;
+            $eventdata->contexturlname   = $info->feedback;
             message_send($eventdata);
         }
     }
diff --git a/mod/feedback/show_nonrespondents.php b/mod/feedback/show_nonrespondents.php
index 4f32247..a2c7f16 100644
--- a/mod/feedback/show_nonrespondents.php
+++ b/mod/feedback/show_nonrespondents.php
@@ -112,7 +112,10 @@ if ($action == 'sendmessage' AND has_capability('moodle/course:bulkmessaging', $
             $eventdata->fullmessage      = html_to_text($htmlmessage);
             $eventdata->fullmessageformat = FORMAT_PLAIN;
             $eventdata->fullmessagehtml  = $htmlmessage;
-            $eventdata->smallmessage     = '';
+            $eventdata->smallmessage     = $subject;
+            $eventdata->courseid         = $course->id;
+            $eventdata->contexturl       = $link3;
+            $eventdata->contexturlname   = $feedback->name;
             $good = $good && message_send($eventdata);
         }
         if (!empty($good)) {
diff --git a/mod/forum/lib.php b/mod/forum/lib.php
index 1966ec1..b231c3d 100644
--- a/mod/forum/lib.php
+++ b/mod/forum/lib.php
@@ -768,6 +768,7 @@ function forum_cron() {
 
                 $eventdata->contexturl = "{$CFG->wwwroot}/mod/forum/discuss.php?d={$discussion->id}#p{$post->id}";
                 $eventdata->contexturlname = $discussion->name;
+                $eventdata->courseid = $course->id;
 
                 $mailresult = message_send($eventdata);
                 if (!$mailresult){
diff --git a/mod/lesson/essay.php b/mod/lesson/essay.php
index c824a63..bb76c61 100644
--- a/mod/lesson/essay.php
+++ b/mod/lesson/essay.php
@@ -219,9 +219,15 @@ switch ($mode) {
                 $a->comment  = s($essayinfo->response);
 
                 // Fetch message HTML and plain text formats
+
+                $b = new stdClass();
+                $b->essay = format_string($pages[$attempt->pageid]->title,true);
+                $b->lesson = $lesson->properties()->name;
+                $b->course = $course->idnumber? $course->idnumber : $course->fullname;
+                $smallmessage  = get_string('essayemailmessage3', 'lesson', $b);
+
                 $message  = get_string('essayemailmessage2', 'lesson', $a);
                 $plaintext = format_text_email($message, FORMAT_HTML);
-
                 // Subject
                 $subject = get_string('essayemailsubject', 'lesson', format_string($pages[$attempt->pageid]->title,true));
 
@@ -233,11 +239,13 @@ switch ($mode) {
                 $eventdata->fullmessage      = $plaintext;
                 $eventdata->fullmessageformat = FORMAT_PLAIN;
                 $eventdata->fullmessagehtml  = $message;
-                $eventdata->smallmessage     = '';
+                $eventdata->smallmessage     = $smallmessage;
+                $eventdata->contexturl       = $CFG->wwwroot . '/grade/report/user/index.php?id=' . $course->id;
 
                 // Required for messaging framework
                 $eventdata->component = 'mod_lesson';
                 $eventdata->name = 'graded_essay';
+                $eventdata->courseid = $course->id;
 
                 message_send($eventdata);
                 $essayinfo->sent = 1;
diff --git a/mod/lesson/lang/en/lesson.php b/mod/lesson/lang/en/lesson.php
index 6dc34e7..c1c4719 100644
--- a/mod/lesson/lang/en/lesson.php
+++ b/mod/lesson/lang/en/lesson.php
@@ -166,6 +166,7 @@ $string['eolstudentoutoftimenoanswers'] = 'You did not answer any questions.  Yo
 $string['essay'] = 'Essay';
 $string['essayemailmessage'] = '<p>Essay prompt:<blockquote>{$a->question}</blockquote></p><p>Your response:<blockquote><em>{$a->response}</em></blockquote></p><p>{$a->teacher}\'s comments:<blockquote><em>{$a->comment}</em></blockquote></p><p>You have received {$a->earned} out of {$a->outof} for this essay question.</p><p>Your grade for the lesson has been changed to {$a->newgrade}&#37;.</p>';
 $string['essayemailmessage2'] = '<p>Essay prompt:<blockquote>{$a->question}</blockquote></p><p>Your response:<blockquote><em>{$a->response}</em></blockquote></p><p>Grader\'s comments:<blockquote><em>{$a->comment}</em></blockquote></p><p>You have received {$a->earned} out of {$a->outof} for this essay question.</p><p>Your grade for the lesson has been changed to {$a->newgrade}&#37;.</p>';
+$string['essayemailmessage3'] = 'Your submission to {$a->lesson} essay {$a->essay} in {$a->course} has been graded.';
 $string['essayemailsubject'] = 'Your grade for {$a} question';
 $string['essays'] = 'Essays';
 $string['essayscore'] = 'Essay score';
diff --git a/mod/quiz/locallib.php b/mod/quiz/locallib.php
index 04ac573..9155a29 100644
--- a/mod/quiz/locallib.php
+++ b/mod/quiz/locallib.php
@@ -1524,6 +1524,7 @@ function quiz_send_confirmation($recipient, $a) {
     $eventdata->smallmessage      = get_string('emailconfirmsmall', 'quiz', $a);
     $eventdata->contexturl        = $a->quizurl;
     $eventdata->contexturlname    = $a->quizname;
+    $eventdata->courseid          = $a->course->id;
 
     // ... and send it.
     return message_send($eventdata);
@@ -1560,6 +1561,7 @@ function quiz_send_notification($recipient, $submitter, $a) {
     $eventdata->smallmessage      = get_string('emailnotifysmall', 'quiz', $a);
     $eventdata->contexturl        = $a->quizreviewurl;
     $eventdata->contexturlname    = $a->quizname;
+    $eventdata->courseid          = $a->course->id;
 
     // ... and send it.
     return message_send($eventdata);
@@ -1611,12 +1613,13 @@ function quiz_send_notification_messages($course, $quiz, $attempt, $context, $cm
     $userstonotify = get_users_by_capability($context, 'mod/quiz:emailnotifysubmission',
             $notifyfields, '', '', '', $groups, $notifyexcludeusers, false, false, true);
 
-    if (empty($userstonotify) && !$sendconfirm) {
+    if (empty($userstonotify) && !$sendconfirm) {die('boo');
         return true; // Nothing to do.
     }
 
     $a = new stdClass();
     // Course info.
+    $a->course          = $course;
     $a->coursename      = $course->fullname;
     $a->courseshortname = $course->shortname;
     // Quiz info.
@@ -1728,6 +1731,7 @@ function quiz_send_overdue_message($course, $quiz, $attempt, $context, $cm) {
     $eventdata->smallmessage      = get_string('emailoverduesmall', 'quiz', $a);
     $eventdata->contexturl        = $a->quizurl;
     $eventdata->contexturlname    = $a->quizname;
+    $eventdata->courseid          = $course->id;
 
     // Send the message.
     return message_send($eventdata);
